crates/compute/src/cpu/layer.rs

```diff
@@ -224,6 +239,107 @@ impl<T: TowerFamily> ComputeLayer<T::B128> for CpuLayer<T> {
 		*accumulator += ret * batch_coeff;
 		Ok(())
 	}
+
+	fn fri_fold<FSub>(
+		&self,
+		exec: &mut Self::Exec,
+		ntt: &impl AdditiveNTT<FSub>,
+		log_len: usize,
+		log_batch_size: usize,
+		challenges: &[T::B128],
+		data_in: &[T::B128],
+		data_out: &mut &mut [T::B128],
+	) -> Result<(), Error>
+	where
+		FSub: binius_field::BinaryField,
+		T::B128: BinaryField + ExtensionField<FSub>,
+	{
+		if data_in.len() != 1 << (log_len + log_batch_size) {
+			return Err(Error::InputValidation(format!(
+				"invalid data_in length: {}",
+				data_in.len()
+			)));
+		}
+
+		if challenges.len() < log_batch_size {
+			return Err(Error::InputValidation(format!(
+				"invalid challenges length: {}",
+				challenges.len()
+			)));
+		}
+
+		if challenges.len() > log_batch_size + log_len {
+			return Err(Error::InputValidation(format!(
+				"challenges length too big: {}",
+				challenges.len()
+			)));
+		}
+
+		if data_out.len() != 1 << (log_len - (challenges.len() - log_batch_size)) {
+			return Err(Error::InputValidation(format!(
+				"invalid data_out length: {}",
+				data_out.len()
+			)));
+		}
+
+		let (interleave_challenges, fold_challenges) = challenges.split_at(log_batch_size);
+		let log_size = fold_challenges.len();
+
+		// Expand the interleave challenges to the tensor size
+		let tensor = self.tensor_expand_with_one(exec, interleave_challenges)?;
+
+		let mut values = vec![T::B128::ZERO; 1 << log_len];
+		for (chunk_index, (chunk, out)) in data_in
+			.chunks_exact(1 << challenges.len())
+			.zip(data_out.iter_mut())
+			.enumerate()
+		{
+			// Fill the values with the folded items
+			if log_batch_size == 0 {
+				values.iter_mut().zip(chunk.iter()).for_each(|(x_i, y_i)| {
+					*x_i = *y_i;
+				});
+			} else {
+				let folded_iter = chunk.chunks(1 << log_batch_size).map(|chunk| {
```

Comment:
We do this method with the tensor fold + inner product because it's somewhat easier with packed fields. However, in this setting of regular field elements, I think it'd be simpler to just do several rounds of 2-to-1 regular folding with linear extrapolations using `interleave_challenges`. In other words, instead of expanding the tensor at all, make the `values` scratchpad have length `1 << (log_len + log_batch_size)`. Copy the whole chunk in and then iterate `extrapolate_line_scalar` over pairs of adjacent elements at the `interleave_challenge`.

This is, I think, a better patter to follow in HW implementation. And it saves 1 multiplication out of 2^log_batch_size.

Reply:
Applied